scripts/biofluid/biofluid.lua

```diff
@@ -39,11 +39,47 @@ local pi = math.pi
 ---@field allocated_fluids_from_providers table<int, number>,
 
 local function migrate_network_data(fluids)
+    local num_migrated = 0
+    local fluids_deleted = {}
+    local MAX_MESSAGE_SIZE = 30
+
     if not fluids then return end
-    for _, key in pairs({"biofluid_robots", "biofluid_requesters"}) do
+    for k, key in pairs({"biofluid_robots", "biofluid_requesters"}) do
         for _, entry in pairs(storage[key]) do
-            entry.name = fluids[entry.name] or entry.name -- if it exists in the table, swap over
-            -- TODO: support deleting fluids completely
+            if fluids[entry.name] == "" then
+                -- This fluid has been deleted from factorio.
+                if key == "biofluid_requesters" then
+                    -- Remove the filter from the requester tank.
+                    entry.name = nil
+                else
+                    -- Delete the in-flight robot and drop the robot item on the ground.
+                    local entity = entry.entity
+                    entity.surface.spill_item_stack {
+                        position = entity.position,
+                        force = entity.force_index,
+                        stack = {name = entity.name, count = 1},
+                        marked_for_deconstruction = true
+                    }
+                    storage.biofluid_robots[k] = nil
```

Comment:
k is the position of `key` in the array {"biofluid_robots", "biofluid_requesters"}, not the robot ID